internal/database/dbconn/open.go

```diff
@@ -69,27 +73,115 @@ func openDBWithStartupWait(cfg *pgx.ConnConfig) (db *sql.DB, err error) {
 	}
 }
 
+// extendedDriver turn a wrapping sql.Driver that doesn't implement Ping,
+// ResetSession and CheckNamedValue into one that does, by reaching for the
+// underlying implementation that actually does support it.
+//
+// A sqlhooks.Driver must be used as a Driver, otherwise this is going to crash.
+type extendedDriver struct {
+	driver.Driver
+}
+type extendedConn struct {
+	driver.ExecerContext
+	driver.QueryerContext
+	driver.Conn
+	driver.ConnPrepareContext
+	driver.ConnBeginTx
+}
+
+// Open returns a conn wrapped through extendedConn, implementing the
+// Ping, ResetSession and CheckNamedValue optional methods that the
+// otelsql.Conn expects to be implemented.
+func (d *extendedDriver) Open(str string) (driver.Conn, error) {
```

Comment:
```suggestion
// otelsql.Conn expects to be implemented.
//
// An error is returned if the underlying driver does not implement
// the required interfaces. TODO
func (d *extendedDriver) Open(str string) (driver.Conn, error) {
```

For the TODO here is to document some hints as to scenarios where we might not fulfill all the required interfaces, to help with debugging etc (I'm not sure I fully understand this myself)

Reply:
@bobheadxi I added a diagram to make this clearer https://github.com/sourcegraph/sourcegraph/pull/39300/files#diff-bc34aa163694ecf12848f338a0eccb4a222a5d61dc30061d597fc78f8238c906R79 

I dislike the idea of adding code that is hard to comprehend, it's bad for maintainability. Does the diagram makes it clearer for you?

Reply:
LGTM 😁 https://github.com/sourcegraph/sourcegraph/pull/39300/files#r955239797

Reply:
Answered positively on Slack :)